src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11302,15 +11320,55 @@ private bool ScanCast(bool forPattern = false)
 
         private bool IsPossibleLambdaExpression(Precedence precedence)
         {
-            // Only call into this if after `static` or after a legal identifier.
+            // Only call into this if after `static`, '[',  or after a legal identifier.
             Debug.Assert(
-                this.CurrentToken.Kind == SyntaxKind.StaticKeyword ||
+                this.CurrentToken.Kind is SyntaxKind.StaticKeyword or SyntaxKind.OpenBracketToken ||
                 this.IsTrueIdentifier(this.CurrentToken));
             if (precedence > Precedence.Lambda)
             {
                 return false;
             }
 
+            if (this.CurrentToken.Kind == SyntaxKind.OpenBracketToken)
+            {
+                var resetPoint = this.GetResetPoint();
+
+                _ = ParseAttributeDeclarations();
+                bool result;
+
+                switch (CurrentToken.Kind)
```

Comment:
I'm not really sure why we can't just unconditionally call `IsPossibleLambdaExpressionCore` here?

Reply:
> I'm not really sure why we can't just unconditionally call `IsPossibleLambdaExpressionCore` here?

`IsPossibleLambdaExpressionCore` is quite complicated and tries to handle a number of cases that could occur after `static` or an identifier, including error recovery. It might also work after `(` for the existing test cases, and after relaxing some asserts, but it seems like a stretch to assume it could handle the same cases unchanged.

---
In reply to: [600876776](https://github.com/dotnet/roslyn/pull/52038#discussion_r600876776) [](ancestors = 600876776)